src/Elastic.Apm/Report/Serialization/LabelsJsonConverter.cs

```diff
@@ -23,16 +23,26 @@ public override void WriteJson(JsonWriter writer, Dictionary<string, string> lab
 					.Replace('"', '_'));
 
 				if (keyValue.Value != null)
-					writer.WriteValue(keyValue.Value.Truncate());
+				{
+					switch (keyValue.Value.Value)
+					{
+						case string strValue:
+							writer.WriteValue(strValue.Truncate());
+							break;
+						default:
+							writer.WriteValue(keyValue.Value.Value);
+							break;
+					}
+				}
 				else
 					writer.WriteNull();
 			}
 			writer.WriteEndObject();
 		}
 
-		public override Dictionary<string, string> ReadJson(JsonReader reader, Type objectType, Dictionary<string, string> existingValue,
+		public override LabelsDictionary ReadJson(JsonReader reader, Type objectType, LabelsDictionary existingValue,
 			bool hasExistingValue, JsonSerializer serializer
 		)
-			=> serializer.Deserialize<Dictionary<string, string>>(reader);
+			=> serializer.Deserialize<LabelsDictionary>(reader);
```

Comment:
This causes a stackoverflow as `Deserialize<LabelsDictionary>` ends up calling the converter's `ReadJson` implementation